File: YourBible/plugins/com-shortsands-utility/src/ios/Utility/Sqlite3.swift
//
//  Sqlite3.swift
//  Utility
//
//  Created by Gary Griswold on 4/19/2019.
//  Copyright © 2018 ShortSands. All rights reserved.
//
// Documentation of the sqlite3 C interface
// https://www.sqlite.org/cintro.html
//

import Foundation
import SQLite3

public enum Sqlite3Error: Error {
    case databaseNotOpenError(name: String)
    case directoryCreateError(name: String, srcError: Error)
    case databaseNotFound(name: String)
    case databaseNotInBundle(name: String)
    case databaseCopyError(name: String, srcError: Error)
    case databaseOpenError(name: String, sqliteError: String)
    case statementPrepareFailed(sql: String, sqliteError: String)
    case statementExecuteFailed(sql: String, sqliteError: String)
}

public class Sqlite3 {
    
    private static var openDatabases = Dictionary<String, Sqlite3>()
    public static func findDB(dbname: String) throws -> Sqlite3 {
        if let openDB: Sqlite3 = openDatabases[dbname] {
            return openDB
        } else {
            throw Sqlite3Error.databaseNotOpenError(name: dbname)
        }
    }
    public static func openDB(dbname: String, copyIfAbsent: Bool) throws -> Sqlite3 {
        if let openDB: Sqlite3 = openDatabases[dbname] {
            return openDB
        } else {
            let newDB = Sqlite3()
            try newDB.open(dbname: dbname, copyIfAbsent: copyIfAbsent)
            openDatabases[dbname] = newDB
            return newDB
        }
    }
    public static func closeDB(dbname: String) {
        if let openDB: Sqlite3 = openDatabases[dbname] {
            openDB.close()
            openDatabases.removeValue(forKey: dbname)
        }
    }
    
    
    private var databaseDir: URL
    private var database: OpaquePointer?
    
    public init() {
        print("****** Init Sqlite3 ******")
        let homeDir: URL = URL(fileURLWithPath: NSHomeDirectory(), isDirectory: true)
        let libDir: URL = homeDir.appendingPathComponent("Library")
        let dbDir: URL = libDir.appendingPathComponent("LocalDatabase")
        
        self.databaseDir = dbDir
        self.database = nil
    }
    // Could introduce alternate init that introduces different databaseDir
    
    deinit {
        print("****** Deinit Sqlite3 ******")
    }
    
    public var isOpen: Bool {
        get {
            return self.database != nil
        }
    }
    
    public func open(dbname: String, copyIfAbsent: Bool) throws {
        self.database = nil
        let fullPath = try self.ensureDatabase(dbname: dbname, copyIfAbsent: copyIfAbsent)
        
        var db: OpaquePointer? = nil
        let result = sqlite3_open(fullPath.path, &db)
        if result == SQLITE_OK {
            print("Successfully opened connection to database at \(fullPath)")
            self.database = db!
        } else {
            print("SQLITE Result Code = \(result)")
            let openMsg = String.init(cString: sqlite3_errmsg(database))
            throw Sqlite3Error.databaseOpenError(name: dbname, sqliteError: openMsg)
        }
    }
    
    /** This open method is used by command line or other programs that open the database at a
     * specified path, not in the bundle.
     */
    public func openLocal(dbname: String) throws {
        self.database = nil
        var db: OpaquePointer? = nil
        let result = sqlite3_open(dbname, &db)
        if result == SQLITE_OK {
            print("Successfully opened connection to database at \(dbname)")
            self.database = db!
        } else {
            print("SQLITE Result Code = \(result)")
            let openMsg = String.init(cString: sqlite3_errmsg(database))
            throw Sqlite3Error.databaseOpenError(name: dbname, sqliteError: openMsg)
        }
    }
    
    private func ensureDatabase(dbname: String, copyIfAbsent: Bool) throws -> URL {
        let fullPath: URL = self.databaseDir.appendingPathComponent(dbname)
        print("Opening Database at \(fullPath.path)")
        if FileManager.default.isReadableFile(atPath: fullPath.path) {
            return fullPath
        } else if !copyIfAbsent {
            try self.ensureDirectory()
            return fullPath
        } else {
            print("Copy Bundle at \(dbname)")
            try self.ensureDirectory()
            let parts = dbname.split(separator: ".")
            let name = String(parts[0])
            let ext = String(parts[1])
            let bundle = Bundle.main
            
            print("bundle \(bundle.bundlePath)")
            let bundlePath = bundle.url(forResource: name, withExtension: ext, subdirectory: "www")
            if bundlePath != nil {
                do {
                    try FileManager.default.copyItem(at: bundlePath!, to: fullPath)
                    return fullPath
                } catch let err {
                    throw Sqlite3Error.databaseCopyError(name: dbname, srcError: err)
                }
            } else {
                throw Sqlite3Error.databaseNotInBundle(name: dbname)
            }
        }
    }
    
    private func ensureDirectory() throws {
        let file = FileManager.default
        if !file.fileExists(atPath: self.databaseDir.path) {
            do {
                try file.createDirectory(at: self.databaseDir, withIntermediateDirectories: true, attributes: nil)
            } catch let err {
                throw Sqlite3Error.directoryCreateError(name: self.databaseDir.path, srcError: err)
            }
        }
    }
    
    public func close() {
        if database != nil {
            sqlite3_close(database)
            database = nil
        }
    }
    
    /**
     * This execute accepts only strings on the understanding that sqlite will convert data into the type
     * that is correct based on the affinity of the type in the database.
     */
    public func executeV1(sql: String, values: [String?], complete: @escaping (_ count: Int) -> Void) throws {
        if database != nil {
            var statement: OpaquePointer? = nil
            let prepareOut = sqlite3_prepare_v2(database, sql, -1, &statement, nil)
            defer { sqlite3_finalize(statement) }
            if prepareOut == SQLITE_OK {
                self.bindStatement(statement: statement!, values: values)
                let stepOut = sqlite3_step(statement)
                if stepOut == SQLITE_DONE {
                    let rowCount = Int(sqlite3_changes(database))
                    complete(rowCount)
                    
                } else {
                    let execMsg = String.init(cString: sqlite3_errmsg(database))
                    throw Sqlite3Error.statementExecuteFailed(sql: sql, sqliteError: execMsg)
                }
            } else {
                let prepareMsg = String.init(cString: sqlite3_errmsg(database))
                throw Sqlite3Error.statementPrepareFailed(sql: sql, sqliteError: prepareMsg)
            }
        } else {
            throw Sqlite3Error.databaseNotFound(name: "unknown")
        }
    }
 
    /**
     * This one is written to conform to the query interface of the cordova sqlite plugin.  It returns
     * a JSON array that can be serialized and sent back to Javascript.  It supports both String and Int
     * results, because that is what are used in the current databases.
     */
    public func queryJS(sql: String, values: [String?], complete: @escaping (_ results: Data) -> Void) throws {
        try queryV0(sql: sql, values: values, complete: { results in
            var message: Data
            do {
                message = try JSONSerialization.data(withJSONObject: results)//,
                                                     //options: JSONSerialization.WritingOptions.prettyPrinted)
            } catch let jsonError {
                print("ERROR while converting resultSet to JSON \(jsonError)")
                let errorMessage = "{\"Error\": \"Sqlite3.queryJS \(jsonError.localizedDescription)\"}"
                message = errorMessage.data(using: String.Encoding.utf8)!
            }
            complete(message)
        })
    }
    
    /**
     * This returns a classic sql result set as an array of dictionaries.  It is probably not a good choice
     * if a large number of rows are returned.  It returns types: String, Int, Double, and nil because JSON
     * will accept these types.
     */
    public func queryV0(sql: String, values: [String?],
                        complete: @escaping (_ results: [Dictionary<String,Any?>]) -> Void) throws {
        if database != nil {
            var resultSet = [Dictionary<String,Any?>]()
            var statement: OpaquePointer? = nil
            let prepareOut = sqlite3_prepare_v2(database, sql, -1, &statement, nil)
            defer { sqlite3_finalize(statement) }
            if prepareOut == SQLITE_OK {
                self.bindStatement(statement: statement!, values: values)
                let colCount = Int(sqlite3_column_count(statement))
                while (sqlite3_step(statement) == SQLITE_ROW) {
                    var row = Dictionary<String, Any?>()
                    for i in 0..<colCount {
                        let col = Int32(i)
                        let name = String(cString: sqlite3_column_name(statement, col))
                        let type: Int32 = sqlite3_column_type(statement, col)
                        if name == "html" {
                            print("Found \(col) \(name) \(type)")
                        }
                        switch type {
                        case 1: // INT
                            row[name] = Int(sqlite3_column_int(statement, col))
                            break
                        case 2: // Double
                            row[name] = sqlite3_column_double(statement, col)
                            break
                        case 3: // TEXT
                            row[name] = String(cString: sqlite3_column_text(statement, col))
                            break
                        case 5: // NULL
                            row[name] = nil
                            break
                        default:
                            row[name] = String(cString: sqlite3_column_text(statement, col))
                            break
                        }
                    }
                    resultSet.append(row)
                }
                complete(resultSet)
            } else {
                let prepareMsg = String.init(cString: sqlite3_errmsg(database))
                throw Sqlite3Error.statementPrepareFailed(sql: sql, sqliteError: prepareMsg)
            }
        } else {
            throw Sqlite3Error.databaseNotFound(name: "unknown")
        }
    }
    
    /**
     * This execute accepts only strings on the understanding that sqlite will convert data into the type
     * that is correct based on the affinity of the type in the database.
     *
     * Also, this query method returns a resultset that is an array of an array of Strings.
     */
    public func queryV1(sql: String, values: [String?], complete: @escaping (_ results:[[String?]]) -> Void) throws {
        if database != nil {
            var resultSet: [[String?]] = []
            var statement: OpaquePointer? = nil
            let prepareOut = sqlite3_prepare_v2(database, sql, -1, &statement, nil)
            defer { sqlite3_finalize(statement) }
            if prepareOut == SQLITE_OK {
                self.bindStatement(statement: statement!, values: values)
                let colCount = Int(sqlite3_column_count(statement))
                while (sqlite3_step(statement) == SQLITE_ROW) {
                    var row: [String?] = [String?] (repeating: nil, count: colCount)
                    for i in 0..<colCount {
                        if let cValue = sqlite3_column_text(statement, Int32(i)) {
                            row[i] = String(cString: cValue)
                        } else {
                            row[i] = nil
                        }
                    }
                    resultSet.append(row)
                }
                complete(resultSet)
            } else {
                let prepareMsg = String.init(cString: sqlite3_errmsg(database))
                throw Sqlite3Error.statementPrepareFailed(sql: sql, sqliteError: prepareMsg)
            }
        } else {
            throw Sqlite3Error.databaseNotFound(name: "unknown")
        }
    }
    
    private func bindStatement(statement: OpaquePointer, values: [String?]?) {
        if let params = values {
            for i in 0..<params.count {
                let col = Int32(i + 1)
                if let param = params[i] {
                    sqlite3_bind_text(statement, col, (param as NSString).utf8String, -1, nil)
                } else {
                    sqlite3_bind_null(statement, col)
                }
            }
        }
    }
    
    public static func errorDescription(error: Error) -> String {
        if error is Sqlite3Error {
            switch error {
            case Sqlite3Error.databaseNotOpenError(let name) :
                return "DatabaseNotOpen: \(name)"
            case Sqlite3Error.directoryCreateError(let name, let srcError) :
                return "DirectoryCreateError \(srcError)  at \(name)"
            case Sqlite3Error.databaseNotFound(let name) :
                return "DatabaseNotFound: \(name)"
            case Sqlite3Error.databaseNotInBundle(let name) :
                return "DatabaseNotInBundle: \(name)"
            case Sqlite3Error.databaseCopyError(let name, let srcError) :
                return "DatabaseCopyError: \(srcError.localizedDescription)  \(name)"
            case Sqlite3Error.databaseOpenError(let name, let sqliteError) :
                return "SqliteOpenError: \(sqliteError)  on database: \(name)"
            case Sqlite3Error.statementPrepareFailed(let sql, let sqliteError) :
                return "StatementPrepareFailed: \(sqliteError)  on stmt: \(sql)"
            case Sqlite3Error.statementExecuteFailed(let sql, let sqliteError) :
                return "StatementExecuteFailed: \(sqliteError)  on stmt: \(sql)"
            default:
                return "Unknown Sqlite3Error"
            }
        } else {
            return "Unknown Error Type"
        }
    }
}

/*
 One person's recommendation to make sqlite threadsafe.
 
 
 +(sqlite3*) getInstance {
 if (instance == NULL) {
 sqlite3_shutdown();
 sqlite3_config(SQLITE_CONFIG_SERIALIZED);
 sqlite3_initialize();
 
 NSLog(@"isThreadSafe %d", sqlite3_threadsafe());
 
 const char *path = [@"./path/to/db/db.sqlite" cStringUsingEncoding:NSUTF8StringEncoding];
 
 if (sqlite3_open_v2(path, &database, SQLITE_OPEN_READWRITE|SQLITE_OPEN_FULLMUTEX, NULL) != SQLITE_OK) {
 NSLog(@"Database opening failed!");
 }
 }
 
 return instance;
 }
 */

File: YourBible/plugins/com-shortsands-utility/src/android/app/src/main/java/com/shortsands/utility/Sqlite3.java
package com.shortsands.utility;

import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteCantOpenDatabaseException;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteException;
import android.util.Log;
import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

/**
 * Created by garygriswold on 4/18/18.
 * Sqlite3.java
 * Utility
 *
 *  Created by Gary Griswold on 4/24/18
 *
 * https://developer.android.com/reference/android/database/sqlite/package-summary.html
 */

public class Sqlite3 {

    private static String TAG = "Sqlite3";

    private static Map<String, Sqlite3> openDatabases = new HashMap<String, Sqlite3>();
    public static Sqlite3 findDB(String dbname) throws SQLiteException {
        Sqlite3 openDB = openDatabases.get(dbname);
        if (openDB != null) {
            return openDB;
        } else {
            throw new SQLiteException("Database " + dbname + " is not open.");
        }
    }
    public static Sqlite3 openDB(Context context, String dbname, boolean copyIfAbsent) throws Exception {
        Sqlite3 openDB = openDatabases.get(dbname);
        if (openDB != null) {
            return openDB;
        } else {
            Sqlite3 newDB = new Sqlite3(context);
            newDB.open(dbname, copyIfAbsent);
            openDatabases.put(dbname, newDB);
            return newDB;
        }
    }
    public static void closeDB(String dbname) {
        Sqlite3 openDB = openDatabases.get(dbname);
        if (openDB != null) {
            openDB.close();
            openDatabases.remove(dbname);
        }
    }


    private Context context;
    private SQLiteDatabase database;

    Sqlite3(Context context) {
        Log.d(TAG,"****** Init AudioSqlite3 ******");
        this.context = context;
        this.database = null;
    }

    public boolean isOpen() {
        return this.database != null && this.database.isOpen();
    }

    public void open(String dbname, boolean copyIfAbsent) throws Exception {
        File fullPath = this.ensureDatabase(dbname, copyIfAbsent);
        int flags = SQLiteDatabase.OPEN_READWRITE |
                SQLiteDatabase.NO_LOCALIZED_COLLATORS |
                SQLiteDatabase.CREATE_IF_NECESSARY;
        this.database = SQLiteDatabase.openDatabase(fullPath.getPath(), null, flags);
        Log.d(TAG,"Successfully opened connection to database at " + fullPath);
    }

    private File ensureDatabase(String dbname, boolean copyIfAbsent) throws IOException {
        File fullPath = this.context.getDatabasePath(dbname);
        Log.d(TAG, "Opening Database as " + fullPath.getAbsolutePath());
        if (fullPath.exists()) {
            return fullPath;
        } else if (! copyIfAbsent) {
            this.ensureDirectory(fullPath);
            return fullPath;
        } else {
            Log.d(TAG,"Copy Bundle at " + fullPath);
            this.ensureDirectory(fullPath);
            try {
                this.copyDatabase(dbname, fullPath);
                return fullPath;
            } catch(IOException ex) {
                throw new IOException("ensureDatabase did not find database in app " + dbname);
            }
        }
    }

    private void ensureDirectory(File destFile) throws IOException {
        String destPath = destFile.getAbsolutePath();
        File destDir = new File(destPath.substring(0, destPath.lastIndexOf("/") + 1));
        try {
            if (!destDir.exists()) {
                destDir.mkdirs();
            }
        } catch(Exception ex) {
            throw new IOException("Could not create " + destDir.getPath());
        }
    }

    private void copyDatabase(String dbName, File destFile) throws IOException {
        InputStream in = null;
        FileOutputStream out = null;
        try {
            in = this.context.getAssets().open("www/" + dbName);
            out = new FileOutputStream(destFile);
            byte[] buf = new byte[1024];
            int len;
            while ((len = in.read(buf)) > 0) {
                out.write(buf, 0, len);
            }
        } finally {
            if (in != null) try { in.close(); } catch(Exception e) {}
            if (out != null) try { out.close(); } catch(Exception e) {}
        }
    }

    public void close() {
        if (this.database != null) {
            this.database.close();
            this.database = null;
        }
    }

    /**
     * This execute is intended for calls from Javascript
     */
    public int executeJS(String sql, JSONArray values) throws Exception {
        if (this.isOpen()) {
            this.database.execSQL(sql, bindJSONArray(values));
            return 1;
        } else {
            throw new SQLiteException("Database is not open.");
        }
    }

    /**
     * This execute is intended for use by other native code
     */
    public int executeV1(String sql, Object[] values) throws SQLiteException {
        if (this.isOpen()) {
            this.database.execSQL(sql, values);
            return 1;
        } else {
            throw new SQLiteException("Database is not open.");
        }
    }

    /**
     * This one is written to conform to the query interface of the cordova sqlite plugin.  It returns
     * a JSON array that can be serialized and sent back to Javascript.  It supports both String and Int
     * results, because that is what are used in the current databases.
     */
    public JSONArray queryJS(String sql, JSONArray values) throws Exception {
        if (this.isOpen()) {
            JSONArray resultSet = new JSONArray();
            Cursor cursor = this.database.rawQuery(sql, bindJSONArray(values));
            int colCount = cursor.getColumnCount();
            int rowNum = 0;
            while(cursor.moveToNext()) {
                JSONObject row = new JSONObject();
                for (int col=0; col<colCount; col++) {
                    String name = cursor.getColumnName(col);
                    int type = cursor.getType(col);
                    switch(type) {
                        case 1: // INT
                            row.put(name, cursor.getInt(col));
                            break;
                        case 2: // Double
                            row.put(name, cursor.getDouble(col));
                            break;
                        case 3: // TEXT
                            row.put(name, cursor.getString(col));
                            break;
                        case 5: // NULL
                            row.put(name, null);
                            break;
                        default:
                            row.put(name, cursor.getString(col));
                            break;
                    }
                }
                resultSet.put(row);
            }
            cursor.close();
            return(resultSet);
        } else {
            throw new SQLiteException("Database is not found.");
        }
    }

    /**
     * This returns a classic sql result set as an array of dictionaries.  It is probably not a good choice
     * if a large number of rows are returned.  It returns types: String, Int, Double, and nil because JSON
     * will accept these types.
     */
    public Cursor queryV0(String sql, Object[] values) throws SQLiteException {
        if (this.isOpen()) {
            return this.database.rawQuery(sql, bindStatement(values));
        } else {
            throw new SQLiteCantOpenDatabaseException("Database must be opended before queryV0.");
        }
    }

    /**
     * This execute accepts only strings on the understanding that sqlite will convert data into the type
     * that is correct based on the affinity of the type in the database.
     *
     * Also, this query method returns a resultset that is an array of an array of Strings.
     */
    public String[][] queryV1(String sql, Object[] values) throws SQLiteException {
        if (this.isOpen()) {
            Cursor cursor = this.database.rawQuery(sql, bindStatement(values));
            int colCount = cursor.getColumnCount();
            String[][] resultSet = new String[cursor.getCount()][colCount];
            int rowNum = 0;
            while (cursor.moveToNext()) {
                String[] row = new String[colCount];
                for (int i = 0; i < colCount; i++) {
                    row[i] = cursor.getString(i);
                }
                resultSet[rowNum] = row;
                rowNum++;
            }
            cursor.close();
            return (resultSet);
        } else {
            throw new SQLiteCantOpenDatabaseException("Database must be opened before query.");
        }
    }

    private String[] bindStatement(Object[] values) {
        String[] result = new String[values.length];
        for (int i=0; i<values.length; i++) {
            result[i] = String.valueOf(values[i]);
        }
        return result;
    }

    private String[] bindJSONArray(JSONArray values) throws JSONException {
        String[] result = new String[values.length()];
        for (int i=0; i<result.length; i++) {
            result[i] = values.getString(i);
        }
        return result;
    }

    public static String errorDescription(Exception error) {
        return ("Sqlite3 " + error.getMessage());
    }
}
